graylog2-shared/src/main/java/org/graylog2/shared/rest/resources/system/SystemPluginResource.java

```diff
@@ -72,15 +49,31 @@ public SystemPluginResource(Set<PluginMetaData> pluginMetaDataSet) {
     @GET
     @Timed
     @ApiOperation(value = "List all installed plugins on this node.")
-    public Map<String, Object> list() {
+    public PluginList list() {
         final List<PluginMetaDataValue> pluginMetaDataValues = Lists.newArrayList();
 
         for (PluginMetaData pluginMetaData : pluginMetaDataSet) {
-            pluginMetaDataValues.add(new PluginMetaDataValue(pluginMetaData));
+            pluginMetaDataValues.add(PluginMetaDataValue.create(
+                    pluginMetaData.getUniqueId(),
+                    pluginMetaData.getName(),
+                    pluginMetaData.getAuthor(),
+                    pluginMetaData.getURL(),
+                    pluginMetaData.getVersion().toString(),
+                    pluginMetaData.getDescription(),
+                    pluginMetaData.getRequiredVersion().toString(),
+                    capabilityToStringSet(pluginMetaData.getRequiredCapabilities())
+            ));
+        }
+
+        return PluginList.create(pluginMetaDataValues);
+    }
+
+    private Set<String> capabilityToStringSet(Set<ServerStatus.Capability> capabilities) {
+        final Set<String> stringSet = Sets.newHashSetWithExpectedSize(capabilities.size());
```

Comment:
An `ImmutableSet` with a fixed size (or even just a view on the original set) would be more efficient here, but since it's only very short-living and the capability set is usually very small, changing that might not make sense. Just a remark.